### ISSUE TITLE
CKEditor - relative URLs not accepted when inserting images.

### ISSUE THREAD

**Issue by Graham-72** [backdrop/backdrop-issues#2220]
The insert image form for a custom block does not seem to accept relative URLs but requires a full URL.

![add custom block image](https://cloud.githubusercontent.com/assets/10559274/18698893/8f654f3e-7fc5-11e6-88f8-a8858049cbba.png)

**Comment by olafgrabienski** [comment]
Related: https://github.com/backdrop/backdrop-issues/issues/2219

**Comment by olafgrabienski** [comment]
@Graham-72 - I was able to use a relative URL of an existing image in a custom block CKEditor dialog. See my second "side note" in https://github.com/backdrop/backdrop-issues/issues/2219#issuecomment-249158882.

**Comment by Graham-72** [comment]
I have now found I have to use a URL like
src="/backdrop/files/inline-images/The_Earth_seen_from_Apollo_17_0.jpg"
but I really want to use
src="/files/inline-images/The_Earth_seen_from_Apollo_17_0.jpg"
or even
src="/inline-images/The_Earth_seen_from_Apollo_17_0.jpg"

**Comment by olafgrabienski** [comment]
@Graham-72 - Are you working on a local installation?

And what looks the URL like:
- http://backdrop.example.dev, or
- http://example.dev/backdrop?

**Comment by Graham-72** [comment]
@olafgrabienski I am working on a VPS. The full URL for the test page is
http://oliverweb.net/justcore/hello-world
and for the image is
http://oliverweb.net/justcore/files/inline-images/The_Earth_seen_from_Apollo_17.jpg
and for its second appearance on that page
http://oliverweb.net/justcore/files/inline-images/The_Earth_seen_from_Apollo_17_0.jpg

I have renamed the backdrop folder to 'justcore' for this test site.

**Comment by olafgrabienski** [comment]
@Graham-72 - Thanks for the URL. I'm not sure but I guess you won't have the path problem if you set up a subdomain for the site: Which would be http://justcore.oliverweb.net instead of http://oliverweb.net/justcore.

**Comment by Graham-72** [comment]
Perhaps related to issue #1292 ?

**Comment by olafgrabienski** [comment]
> Perhaps related to issue #1292 ?

@Graham-72 - I don't think so, because in that issue a path like `/backdrop/files/...` didn't work while in https://github.com/backdrop/backdrop-issues/issues/2220#issuecomment-249192177 you said, you was able to use `/backdrop/files/...`.

Did you manage to set up a subdomain as recommended in https://github.com/backdrop/backdrop-issues/issues/2220#issuecomment-249198917? That should work.

Another approach could be to set a Base URL (cf. _settings.php_, line 73 ff.), but I'm short on experience with using it. If you want to try it, you might consider to have a look at the RewriteBase in the _.htaccess_ file, line 100 ff, as well.

**Comment by Graham-72** [comment]
@olafgrabienski thank you for your comments on this. Actually I have not tried setting up a subdomain and do not presently want to do this for this test site.

Maybe this feature in Backdrop is working as intended. In which case my only issue is that the 'hint' /example/image.jpg does not fully indicate which relative URL is needed here.

**Comment by olafgrabienski** [comment]
@Graham-72 - When you don't want to use a subdomain, I recommend to test the Base URL approach as mentioned in https://github.com/backdrop/backdrop-issues/issues/2220#issuecomment-249393612.

Because using a path like `/backdrop/files/...` in such dialogs, you get into trouble when you rename the `backdrop` folder, for instance when you transfer the site to another environment.

(Can anyone confirm that a Base URL in *settings.php* makes sense in this case, allowing to use relative URLs like `/files/inline-images/my-image.png` instead of `/backdrop/files/inline-images/my-image.png` in a situation as described in the issue description, cf. https://github.com/backdrop/backdrop-issues/issues/2220#issue-178251060?)

**Comment by docwilmot** [comment]
I recently found a persistently failing test on my local WAMP, and discovered this was the reason. Subdirectory  installs prepend the directory name (like `/backdrop/`) to the base URL, and no, fiddling with `settings.php` did not help this. Checked extensively on Drupal.org (because I really couldnt bother to figure out subdomains on WAMP - turns out its a one-liner to do though) and it seems that this is expected behavior.

So in short, since a user would probably expect Backdrop to work in a subdirectory, this is a bug, and should be either fixed or explained.

**Comment by olafgrabienski** [comment]
Unfortunately, I either managed to figure this issue out on Drupal. That's why I went over to use subdomains or domains without subdirectory path. However, once I *had to* work with a subdirectory path, which caused a lot of trouble with some images. It would be great, if we could fix that in Backdrop!

(Apart from that, re WAMP: for local Backdrop installations, I use AMPPS and, for a short time, Kalabox. In AMPPS it's easy to set up domains for your projects; Kalabox generates them automatically during site install.)

**Comment by Graham-72** [comment]
> you get into trouble when you rename the backdrop folder, for instance when you transfer the site to another environment

It is my understanding that the full URL is stored in the Backdrop database (field_data_body / body_value) and one has to accept that editing is needed if the Backdrop directory is changed. Does using a subdomain change this? What would be stored in the database?

**Comment by olafgrabienski** [comment]
My understanding is: When you use a domain or a subdomain (i.e. without a backdrop subdirectory in the path) and declare your image as `files/inline-images/my-image.jpg`, no absolute path would be stored in the database, assuming the file is in the 'base' directory of Backdrop CMS, and irrespective of the directory's name. Consequently you will be able to move your installation easily, without editing URLs or image paths. @docwilmot, can you confirm my assumption?

**Comment by jenlampton** [comment]
If what is inserted into the body area is an image tag with data attributes containing a file ID -- as it should on Backdrop sites -- the actual path to the file is only important at render time. It doesn't matter if the location of the file changes between when the file was inserted and when it was rendered, or if the domain, sub-domain, directory, or anything else in the path changes. As long as the location of the file can be properly deduced when the page is viewed, this should be a non-issue for Backdrop sites. (In Drupal sites, a relative, or worse, absolute path was stored in the body field).

Can whoever has been experiencing this problem determine what is stored in the field? If it's not an image tag with data attributes containing a file ID we should figure out why.

**Comment by quicksketch** [comment]
> If what is inserted into the body area is an image tag with data attributes containing a file ID -- as it should on Backdrop sites -- the actual path to the file is only important at render time. It doesn't matter if the location of the file changes between when the file was inserted and when it was rendered, or if the domain, sub-domain, directory, or anything else in the path changes.

I don't think any of this statement is true. We don't do any parsing of file IDs on output. We parse file IDs on *save*, but on output the markup for `<img>` tags is unaffected by the `data-file-id` attribute.

Overall, Backdrop is doing what is expected here. In the case of a manually-specified path, there's not going to be a File ID in the first place, it's just an arbitrary file with a path. The value that is typed into that field is exactly the value that will be put into the `<img>` tag's `src` attribute.

In order to accommodate for the site directory being moved around, Backdrop would need to port the [Pathologic module](https://www.drupal.org/project/pathologic). But in general, I've always found it a lot easier to just always set up a domain (or matching directory) on all servers where I have a copy of the same site so that the relative path is always the same.

**Comment by jenlampton** [comment]
> I don't think any of this statement is true. We don't do any parsing of file IDs on output. We parse file IDs on save, but on output the markup for <img> tags is unaffected by the data-file-id attribute.

Ugh. So we still have all Drupal's problems? :(

**Comment by quicksketch** [comment]
In this case, yes. Especially because here we're talking about an arbitrary file path. There's no upload or File ID happening here, so all Backdrop has to go on is the path that is provided by the user.

**Comment by Graham-72** [comment]
> Can whoever has been experiencing this problem determine what is stored in the field

It would seem that for custom blocks it is stored in the config file for the block and not in the database. What is stored is whatever is entered, so it must be a meaningful relative URL.

Is there possibly a way that Backdrop might help the user here by showing the current base address for the site?

Now that I understand exactly what is happening I shall be OK in future. In the past on Drupal I have found Pathologic useful and we have it as a contrib module for Backdrop. Perhaps one day some of its cleverness might be added to core?

**Comment by jenlampton** [comment]
> Is there possibly a way that Backdrop might help the user here by showing the current base address for the site?

This is a good idea! Perhaps we can add it as a `field_prefix` when someone is entering a relative URL? Or, do we not currently differentiate between on-site and off-site image sources in the UI? I wonder if separating those two types of input in the UI would help or hinder.... something like

```
[ ] Internal Image
http://www.actualsite.com/subdir/[_example.png_______________]
```
vs 
```
[ ] External Image
[_http://www.examplesite.com/example.png________]
```

**Comment by quicksketch** [comment]
>  Or, do we not currently differentiate between on-site and off-site image sources in the UI?

We don't ask or know about this information currently. I don't think it would be appreciated making users fill out another form field just so we could update the UI to match. The vanilla CKEditor does something automatic with URLs via JavaScript, but I've found its behavior to be more confusing rather than helpful. You can see what it does in the CKEditor demo page (insert a link): http://ckeditor.com/demo

**Comment by Graham-72** [comment]
> Can anyone confirm that a Base URL in settings.php makes sense in this case, allowing to use relative URLs like /files/inline-images/my-image.png

I am finding that this does not work and Backdrop is not applying the Base URL to URLs of images added by CKEditor. With the change introduced at #2261 I can no longer use the image browse facility in CKEditor to add or edit images in a site that is in a subdirectory. Changing the RewriteBase in .htaccess does not seem to solve this problem either.

Surely this is a bug in the system?

**Comment by Graham-72** [comment]
I have now worked out what my problem has been, and found a solution.

If Backdrop is installed in a subdirectory one will normally include an .htaccess file in the site's root directory that will have 'RewriteRule ^(.*)$ backdrop/$1 [L]' where 'backdrop' is the subdirectory name.

 This rewrite rule enables the /files/inline-images folder to be correctly found and used.

But if for some reason you wish to use a Backdrop installation in a subdirectory independently of the rest of the site, perhaps as a transition stage of moving from one version of the site to another, one would not have this rewrite rule in the root folder but must add a more specialised one to the htaccess file in the Backdrop subdirectory.

I am far from expert at use of htaccess but here is a rewrite rule that seems to work
RewriteRule ^files/inline-images/(.+)$ http://www.mydomain.com/mysubdirectory/files/inline-images/$1

If anyone can suggest something better, I will be delighted to hear it.

**Comment by Graham-72** [comment]
Thanks to the change introduced by issue #2261 I can now reliably and confidently use CKEditor to add images to nodes and blocks even though my site is installed in a subdirectory, for example 'backdrop'. The settings I use in Backdrop to achieve this are listed below. When installing a new site it would seem best to turn off caching and bandwidth optimization whilst making these changes. 

1. In /backdrop/settings.php uncomment the Base URL statement (currently at line 91) and fill in the domain name for the Backdrop installation e.g. $base_url = 'http://www.mydomain.com';

2. In /backdrop/.htaccess change the line 'RewriteRule ^ index.php [L]' (currently line 122) to be 'RewriteRule ^(.*)$ index.php?q=$1 [L,QSA]'. _(I do not know enough to explain why this is necessary!)_

3. Also in /backdrop/.htaccess uncomment line 104 to set the rewrite base 'RewriteBase /backdrop'

4. Add the following lines in a new .htaccess file in the base directory i.e. /.htaccess
Options +FollowSymlinks
RewriteEngine on
RewriteRule ^/?$ index.php [L]
RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteCond %{REQUEST_URI} !=/favicon.ico
RewriteRule ^(.*)$ backdrop/$1 [L]

Warning: I am no expert on this, and I do not understand all the why's and wherefore's. However, it seems to do what is needed on my server. Perhaps an expert explanation will be forthcoming, maybe even a better solution?

**Comment by olafgrabienski** [comment]
Phew, looks tricky! Sadly I'm not an expert of rewrite rules at all. Good that they work for you! It's a workaround though, and I'd wish that users could install Backdrop in a subdirectory without the need for such special care (should probably be a separate issue).

**Comment by quicksketch** [comment]
It sounds like the original request (relative URLs in CKEditor) has been fixed, with a combination of bug fixes in Backdrop plus #1613, which actually uses root-relative URLs by default now.

Regarding the sub-directory installation of Backdrop, it may vary based on server. The default .htaccess and rules should work without modification on most servers. I personally still haven't been able to establish when `RewriteBase` is needed and when it's not. But yes, definitely separate from the CKEditor question.
